File: ALG_HW8/src/ALG_HW8_Task3.java
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ALG_HW8_Task3 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int nInput;
        String buffer;
        List<Node> nodeInfo;
        String errorMsg = "not complete";

        nInput = Integer.parseInt(input.nextLine());
        input.nextLine();
        for (int i = 0; i < nInput; i++) {
            buffer = input.nextLine();
            nodeInfo = getNodeList(buffer);
            if (!nodeInfo.isEmpty()) {
                if (validateTree(nodeInfo)) treeLevelTraversal(nodeInfo);
                else {
                    System.out.println(errorMsg);
                }
            } else {
                System.out.println(errorMsg);
            }
        }
    }

    public static boolean validateStringFormat(String text, String format) {
        Pattern pattern = Pattern.compile(format);
        Matcher matcher = pattern.matcher(text);

        return matcher.matches();
    }

    public static Node getNodeInfo(String infoBuffer) {
        String[] info = infoBuffer.substring(1, infoBuffer.length() - 1).split(",");
        if (validateStringFormat(infoBuffer, "\\(\\d+,\\)")) {
            return new Node(Integer.parseInt(info[0]), "");
        }
        return new Node(Integer.parseInt(info[0]), info[1]);
    }

    public static List<Node> getNodeList(String infoBuffer) {
        HashMap<String, Node> cache = new HashMap<>();
        ArrayList<Node> nodeInfo = new ArrayList<>();
        String[] infos = infoBuffer.split(" ");
        Node nodeBuffer;

        if (!validateStringFormat(infoBuffer, "(\\(\\d+,[LR]*\\) )+\\(\\)")) return Collections.emptyList();
        for (String s : infos) {
            if (s.equals("()")) break;
            if (validateStringFormat(s, "\\(\\d+,[LR]*\\)")) {
                nodeBuffer = getNodeInfo(s);
                if (cache.containsKey(nodeBuffer.getPath())) {
                    return Collections.emptyList();
                }
                cache.put(nodeBuffer.getPath(), nodeBuffer);
            } else {
                return Collections.emptyList();
            }
        }
        cache.forEach((k, v) -> nodeInfo.add(v));
        nodeInfo.sort(Node.compare);

        return nodeInfo.get(0).getPath().equals("") ? nodeInfo : Collections.emptyList();
    }

    public static boolean validateTree(List<Node> nodeInfo) {
        List<Node> nodes = new ArrayList<>(nodeInfo);
        Node root = nodes.get(0);
        Node cur;
        Node next;
        String pathBuffer;
        int cnt = 1;

        if (nodeInfo.isEmpty()) return false;
        nodes.remove(0);
        nodes.sort(Node.compare);
        for (Node n : nodes) {
            pathBuffer = n.getPath();
            cur = root;
            for (int i = 0; i < pathBuffer.length(); i++) {
                if (i == pathBuffer.length() - 1) {
                    if (pathBuffer.charAt(i) == 'L') {
                        cur.setLeft(n);
                    } else {
                        cur.setRight(n);
                    }
                    ++cnt;
                    break;
                }
                next = (pathBuffer.charAt(i) == 'L') ? cur.getLeft() : cur.getRight();
                if (next == null) return false;
                cur = next;
            }
        }
        return cnt == nodeInfo.size();
    }

    public static void treeLevelTraversal(List<Node> nodeInfo) {
        nodeInfo.sort(Node.compareLevel);
        for (Node n : nodeInfo) System.out.print(n.getValue() + " ");
        System.out.println();
    }

    static class Node {
        static Comparator<Node> compare = Comparator.comparing(Node::getPath);
        static Comparator<Node> compareLevel = (n1, n2) -> {
            int pathLen1 = n1.getPath().length();
            int pathLen2 = n2.getPath().length();
            if (pathLen1 == pathLen2) return n1.getPath().compareTo(n2.getPath());
            else {
                return pathLen1 < pathLen2 ? -1 : 1;
            }
        };
        private final int value;
        private final String path;
        private Node left;
        private Node right;

        public Node(int value, String path) {
            this.value = value;
            this.path = path;
            left = null;
            right = null;
        }

        int getValue() {
            return value;
        }

        String getPath() {
            return path;
        }

        Node getLeft() {
            return left;
        }

        void setLeft(Node node) {
            left = node;
        }

        Node getRight() {
            return right;
        }

        void setRight(Node node) {
            right = node;
        }
    }
}

File: ALG_HW8_Task3.cpp
#include <iostream>
#include <string>
#include <regex>
#include <utility>
#include <vector>
#include <algorithm>

using namespace std;

class Node {
private:
    int value;
    string path;
    Node *left{};
    Node *right{};

public:
    explicit Node(int v = 0, string  p = "", Node *n1 = nullptr, Node* n2 = nullptr):
    value(v), path(std::move(p)), left(n1), right(n2) {};

    [[nodiscard]] int getValue() const;

    [[nodiscard]] string getPath() const;

    Node *getLeft();

    void setLeft(Node *node);

    Node *getRight();

    void setRight(Node *node);

};

Node *getNodeInfo(const string &nodeInfo);

bool findNode(const vector<Node *> &nodeList, const string &path);

vector<Node *> getNodeList();

bool compareNode(Node *n1, Node *n2);

bool compareNodeLevel(Node *n1, Node *n2);

bool validateTree(vector<Node *> nodeList);

void treeLevelTraversal(vector<Node *> nodeList);

int main() {
    int nInput;
    vector<Node *> nodeInfoBuffer;
    regex inputPattern(R"((\(\d+,[LR]*\)))");

    cin >> nInput;
    for (int i = 0; i < nInput; i++) {
        nodeInfoBuffer = getNodeList();
        if (!nodeInfoBuffer.empty()) {
            if (validateTree(nodeInfoBuffer)) treeLevelTraversal(nodeInfoBuffer);
            else cout << "not complete" << endl;
        } else cout << "not complete" << endl;
    }

    return 0;
}

int Node::getValue() const { return value; }

string Node::getPath() const { return path; }

Node *Node::getLeft() { return left; }

void Node::setLeft(Node *node) { left = node; }

Node *Node::getRight() { return right; }

void Node::setRight(Node *node) { right = node; }

Node *getNodeInfo(const string &nodeInfo) {
    unsigned long splitPoint = nodeInfo.find(',');
    unsigned long length = nodeInfo.length();
    int value = stoi(nodeInfo.substr(1, splitPoint - 1));
    string path = nodeInfo.substr(splitPoint + 1, length - splitPoint - 2);

    if (regex nodePattern(R"(\(\d+,\))"); !regex_match(nodeInfo, nodePattern)) return new Node(value, path);
    return new Node(value);
}

bool findNode(const vector<Node *> &nodeList, const string &path) {
    return any_of(nodeList.begin(), nodeList.end(), [path](Node const *n) { return n->getPath() == path; });
}

vector<Node *> getNodeList() {
    string inputBuffer;
    regex inputPattern(R"((\(\d+,[LR]*\)))");
    Node *nodeBuffer;
    vector<Node *> nodeList;

    while (true) {
        cin >> inputBuffer;
        if (inputBuffer == "()") break;
        if (regex_match(inputBuffer, inputPattern)) {
            nodeBuffer = getNodeInfo(inputBuffer);
            if (findNode(nodeList, nodeBuffer->getPath())) return {};
            nodeList.push_back(nodeBuffer);
        } else return {};
    }
    sort(nodeList.begin(), nodeList.end(), compareNode);

    if (nodeList.empty()) return {};
    if (!nodeList[0]->getPath().empty()) return {};
    return nodeList;
}

bool compareNode(Node *n1, Node *n2) { return n1->getPath() < n2->getPath(); }

bool compareNodeLevel(Node *n1, Node *n2) {
    unsigned long pathLength1 = n1->getPath().length();
    unsigned long pathLength2 = n2->getPath().length();
    if (pathLength1 == pathLength2) return compareNode(n1, n2);
    else return pathLength1 < pathLength2;
}

bool validateTree(vector<Node *> nodeList) {
    vector<Node *> nodes;
    Node *root = nodeList[0];
    Node *cur;
    Node *next;
    string pathBuffer;
    int cnt = 1;

    if (nodeList.empty()) return false;
    for (int i = 1; i < nodeList.size(); i++) nodes.push_back(nodeList[i]);
    sort(nodes.begin(), nodes.end(), compareNode);
    for (Node *n: nodes) {
        pathBuffer = n->getPath();
        cur = root;
        for (int i = 0; i < pathBuffer.length() - 1; i++) {
            next = (pathBuffer[i] == 'L') ? cur->getLeft() : cur->getRight();
            if (next != nullptr) cur = next;
            else {
                nodes.clear();
                return false;
            }
        }
        if (pathBuffer[pathBuffer.length() - 1] == 'L') cur->setLeft(n);
        else cur->setRight(n);
        ++cnt;
    }
    nodes.clear();
    return cnt == nodeList.size();
}

void treeLevelTraversal(vector<Node *> nodeList) {
    sort(nodeList.begin(), nodeList.end(), compareNodeLevel);
    cout << nodeList[0]->getValue();
    for (int i = 1; i < nodeList.size(); i++) cout << " " << nodeList[i]->getValue();
    cout << endl;
}